### PR TITLE
[logging] Superficial use of new metrics_context in CompilationMetrics logging

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139263
* __->__ #139260
* #139259

Summary: This change makes use of the new metrics accounting utility to record CompilationMetrics. This is a first step and not a deep integration: wherever we were previously creating a CompilationMetrics object to log, we instead update the current metrics_context and rely on that contextmanager to call record_compilation_metrics when it exits. In a subsequent change, we'll add usage of the 'timed' contextmanager to collect various compile times.

Test Plan: unit test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by masnesral** [comment]
Looks like many (or all?) of the test failures are due to my check that we don't re-enter the contextmanager again before exiting. So either I've put the contextmanager on the backward "too low" or there actually is good reason to allow re-entering. Any insights? I can update the code to support logging on exit of the outermost contextmanager if that's needed.

**Comment by ezyang** [comment]
Re the reentrant problem, my opinion is that you should transparently no-op the inner context when this occurs.

Here is my reasoning: the raison d'etre of dynamo_compile is we want an event for every top level invocation of compilation, *no matter* where it happens. If somehow, I end up calling into inductor compilation without having already having a dynamo context, I *still* want to generate a dynamo_compile event. This is similar to the discussion we had about how to SUM compile time in pt2_compile_events: you simply take every event that has an empty stack... and they may NOT all be the same events.

The de facto compilation metrics spans are Dynamo compile, and AOTAutograd backwards (sort of; the actual CompilationMetrics is currently triggered from the dynamo_timed on the Inductor itself lol). We identified in our recent meeting one more event that's missing (runtime autotuning, which needs to be done specially since it's interleaved with eager code). But there may be other compile sites we are missing that ought to be represented. In fact, the most reliable way to ensure you have an event everywhere is to whole hog put *everything* into dynamo_timed LOL. I'm OK if you don't go that far on this stack though.

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -810,6 +811,11 @@ def to_int_us(v: Optional[float]) -> Optional[int]:
     return None if v is None else int(v * 1_000_000)
 
 
+# Version field added to every log. Increment to make it easier to distinguish new
+# vs. old entries when you make a substantive change to how the logs are populated.
+LOG_FORMAT_VERSION = 1
```
Comment (ezyang) [inline torch/_dynamo/utils.py:816]: Do this... separately, maybe? lol

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -954,6 +959,33 @@ def get_compilation_metrics() -> List[CompilationMetrics]:
...
+# To record execution time:
+#    def foo(...):
+#        # Record microseconds
+#        with metrics_context.timed("metric_us")
+#            ...
+#        # Record milliseconds
+#        with metrics_context.timed("metric_ms")
+#            ...
+#
+metrics_context = MetricsContext(on_exit=record_compilation_metrics)
```
Comment (ezyang) [inline torch/_dynamo/utils.py:986]: nit: make the global all caps

**Pull request closed without merging**